dashboard/app/controllers/script_levels_controller.rb

```diff
@@ -247,15 +247,37 @@ def stage_extras
       return
     end
 
-    @stage = Script.get_from_cache(params[:script_id]).lesson_by_relative_position(params[:lesson_position].to_i)
+    @stage = Script.get_from_cache(
+      params[:script_id]
+    ).lesson_by_relative_position(
+      params[:lesson_position].to_i
+      )
     @script = @stage.script
+    script_bonus_levels_by_stage = @script.get_bonus_script_levels(@stage)
+
+    user = @user || current_user
+    unless user.nil?
+      # bonus level summaries explicitly don't contain any user-specific data,
+      # so we need to merge in the user's progress.
+      script_bonus_levels_by_stage.each do |stage|
```

Comment:
I'm curious, why not pass the user to `get_bonus_script_levels` and do the merging there?